---
title: ""
output: html_document
---

```{r setup, include=FALSE}
library(magrittr)
library(tidyverse)
library(vroom)

library(DESeq2)
library(ggdendro)

knitr::opts_chunk$set(echo = TRUE)
```

## 2021-07-21: RNA-Seq Clustering and Differential Expression Testing

Please email me your Rmarkdown document at <kkeith@coriell.org> with your **LAST NAME** in the filename by 9AM Friday 7/21.

---

### RNA-seq with Full Data Set

#### Read / Wrangle Data

Taking the data in the zipped folder `full_count_files` which contains the full RNA-seq count tables for the human genome (not just the chr21 files we were using for the demo) and use the chunk below to read it in, and convert it into a filtered `DESeq` object. (Basically you should just copy/paste your code in the chunk below from yesterday's assignment). Make sure to specify which treatment is the control/reference treatment.

```{r}
### read data in here
# get list of files
list.files(path = '../full_count_files',
           pattern = '.txt$',
           full.names = T) -> files

### read files in recursively with vroom()
vroom::vroom(files, id = 'file_path', delim = '\t', comment = '#', skip = 2, 
             col_names = c('ensembl_gene_id', 'chr', 'start', 'end', 
                           'strand', 'length', 'count')) %>%
  mutate(sample_id = str_extract(file_path, '[d,s][A-z]{2,3}[0-9]'),
         treatment = str_remove_all(sample_id, '[0-9]')) %>%
  dplyr::select(-file_path) -> data

# transform the data table into a count matrix
data %>%
  dplyr::select(ensembl_gene_id, sample_id, count) %>%
  spread(sample_id, count) %>% 
  as.data.frame() -> count_matrix

# get a metadata table
data %>%
  dplyr::select(sample_id, treatment) %>%
  distinct() -> metadata

# give count matrix and metadata table and transform into a DESeq object
dds <- DESeqDataSetFromMatrix(countData = count_matrix,
                              colData = metadata,
                              tidy = TRUE,
                              design = ~ treatment)
### extra argument - make sure DESeq knows what the reference is 
dds$treatment <- relevel(dds$treatment, ref = "siC")

### filter out rows that contain only zero counts
keep <- rowSums(counts(dds)) >= 1
dds <- dds[keep, ]

# Filter for genes where at least three samples have a count of 10 or higher
keep <- rowSums(counts(dds) >= 10) >= 3
dds <- dds[keep, ]

### OPTIONAL, but recommended: check how many genes are filtered out
table(keep)
```

<br>

### Check for Treatment and Batch Effects by Clustering

#### Normalize the Data

```{r}
# have to calculate normalization factors first
dds <- estimateSizeFactors(dds)

# get the normalized counts
fpm(dds) %>%
  as.data.frame() %>%
  rownames_to_column('ensembl_gene_id') -> norm_counts
```

#### Hierarchical Clustering

Use the chunk below to plot a dendrogram.

```{r}
### calculate the distances and clustering for the tree
norm_counts %>% 
  column_to_rownames('ensembl_gene_id') %>%
  t() %>%
  dist(.) %>% 
  fastcluster::hclust() %>%
  as.dendrogram() %>%
  dendro_data() -> clust_dds

### can just use ggdendrogram
ggdendrogram(clust_dds)

### plot the most pretty with ggplot
# add treatment information to the labels for the best coloring
clust_dds$labels %>% 
  mutate(treatment = str_remove_all(label, '[0-9]'),
         y = -25) -> clust_labels

# plot with ggplot2
ggplot() +
  geom_segment(data = clust_dds$segments, 
               aes(x = x, y = y, xend = xend, yend = yend)) +
  geom_text(data = clust_labels, 
            aes(x = x, y = y, label = label, color = treatment), 
            vjust = 1.2, size = 6) +
  scale_color_manual(values = c('gold3', 'hotpink3', 'cyan4')) +
  coord_cartesian(ylim = c(-100, max(clust_dds$segments$yend))) +
  theme_classic() +
  theme(legend.position = 'none',
        axis.title = element_blank(), 
        axis.line = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank())
```

#### PCA

Use the chunk below to plot a PCA

```{r}
### using DESeq functions
# normalize the data
vsd <- varianceStabilizingTransformation(dds, blind = FALSE)

# plot
plotPCA(vsd, intgroup = 'treatment') +
  geom_point(size = 6) +
  scale_color_manual(values = c('gold3', 'hotpink4', 'cyan4')) +
  theme_classic()

### OR doing it yourself
# Calculate the PCA
norm_counts %>%
  column_to_rownames('ensembl_gene_id') %>%
  t() %>%
  prcomp() -> pca

# plot
pca$x %>%
  as.data.frame() %>%
  rownames_to_column('sample_id') %>%
  mutate(treatment = str_remove_all(sample_id, '[0-9]')) %>%

ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(color = treatment), size = 6) +
  scale_color_manual(values = c('gold3', 'hotpink4', 'cyan4')) +
  theme_classic(base_size = 16) +
  theme(legend.position = 'top')
```
<br>

---

<br>

### Run Differential Expression for 2 Different Treatments

#### Test

Use the chunk below to run differential expression

```{r}
dds <- DESeq(dds)
```

#### Get and Save Results

Use the chunk below to extract the differential expression results for both treatments and save the tables.

```{r}
# SF3B1 Inhibition
results(dds, contrast = c('treatment', 'siSF', 'siC')) %>%
  as.data.frame() %>%
  rownames_to_column('ensembl_gene_id') %>%
  mutate(treatment = 'siSF') %>%
  select(treatment, everything()) -> siSF_results

# DAC treatment
results(dds, contrast = c('treatment', 'dac', 'siC')) %>%
  as.data.frame() %>%
  rownames_to_column('ensembl_gene_id') %>%
  mutate(treatment = 'dac') %>%
  select(treatment, everything()) -> dac_results

# write the results
rbind(siSF_results, dac_results) %>%
  write_tsv('diff_exp.tsv')
```





